python/src/iceberg/table/schema.py

```diff
@@ -0,0 +1,179 @@
+                    return field.field_id
+        raise ValueError(f"Cannot get field ID, name not found: {field_identifier}")
+
+    def _get_field_identifier_by_alias(self, field_identifier: str, case_sensitive: bool) -> int:
+        """Get a field ID for a given field alias
+
+        Args:
+            field_identifier (str): A field alias
+            case_sensitive (bool): If False, case will not be considered when retrieving the field from the schema
+
+        Returns:
+            int: The field ID for the field alias
+        """
+        if case_sensitive:
+            try:
+                # For case-sensitive, just try looking up the alias and raise if it's not found
+                return self._aliases[field_identifier]
+            except KeyError:
+                raise ValueError(f"Cannot get field ID, alias not found: {field_identifier}")
+        if not case_sensitive:
+            matching_fields = [value for key, value in self._aliases.items() if key.lower() == field_identifier.lower()]
+            if len(matching_fields) == 1:  # If one matching alias found, return the corresponding ID
+                return matching_fields[0]
+            elif not matching_fields:  # If no matching fields, raise
+                raise ValueError(f"Cannot get field ID, alias not found: {field_identifier}")
+
+            # If multiple IDs are returned for a case-insensitive alias lookup, raise
+            raise ValueError(f"Cannot get field ID, case-insensitive alias returns multiple results: {field_identifier}")
+
+    def get_field_id(
+        self,
+        field_identifier: str,
+        field_identifier_type: FIELD_IDENTIFIER_TYPES = FIELD_IDENTIFIER_TYPES.NAME,
+        case_sensitive: bool = True,
+    ) -> int:
+        """Get a field ID for a given NAME or ALIAS
+
+        This calls either the `_get_field_identifier_by_name` method or the `_get_field_identifier_by_alias` method depending on the value of the `field_identifier_type` argument.
+
+        Args:
+            field_identifier (str): The unique identifier for the field
+            field_identifier_type (FIELD_IDENTIFIER_TYPES): An FIELD_IDENTIFIER_TYPES value of either NAME or ALIAS
+            case_sensitive (bool): If False, case will not be considered when retrieving the field from the schema. Default is True
+
+        Raises:
+            ValueError: If the field identifier does not exist in the schema
+
+        Returns
+            int: The field ID for the given field identifier
+        """
+        if field_identifier_type == FIELD_IDENTIFIER_TYPES.NAME:
+            field_id = self._get_field_identifier_by_name(field_identifier=field_identifier, case_sensitive=case_sensitive)
+        elif field_identifier_type == FIELD_IDENTIFIER_TYPES.ALIAS:
+            field_id = self._get_field_identifier_by_alias(field_identifier=field_identifier, case_sensitive=case_sensitive)
+
+        return field_id
+
+    def get_field(self, field_id: int) -> NestedField:
```

Comment:
These should use maps that are constructed by indexing with a SchemaVisitor. All of the methods on schema are intended to return a field by full name or by ID from anywhere in the schema. That's why schema is not just a regular struct. Structs will return fields by name or ID, but are limited to just that struct.

That's also why we use names like `find_field` rather than `get_field` (in addition, we avoid `get` in methods names generally).

Reply:
Thanks for the explanation, I saw that in the legacy implementation but now I understand the purpose. I'll update this today.

Reply:
@rdblue I updated the PR to use the visitor pattern and added the `IndexById` and `IndexByName` schema visitors.